File: deployment-services/infrastructure-service/api/serializers/infrastructure.py
from typing import Any

from api.models.infrastructure import Infrastructure
from api.types.infrastructure import InfrastructureResponse


class InfrastructureSerializer:
    @staticmethod
    def serialize_instance(instance: Infrastructure) -> dict[str, Any]:
        invited_users_details = [
            {
                'id': str(u.id),
                'email': u.email,
                'user_name': u.user_name,
                'role': u.role,
            }
            for u in instance.invited_users.all()
        ]
        try:
            env = instance.environments.latest('created_at')
            status = env.status
        except Exception:
            status = "UNKNOWN"

        response = InfrastructureResponse(
            id=instance.id,
            name=instance.name,
            user_id=instance.user_id,
            cloud_provider=instance.cloud_provider,
            max_cpu=instance.max_cpu,
            max_memory=instance.max_memory,
            is_cloud_authenticated=instance.is_cloud_authenticated,
            code=instance.code,
            metadata=instance.metadata,
            created_at=instance.created_at,
            updated_at=instance.updated_at,
            invited_users=invited_users_details,
            status=status,
            is_mock=instance.is_mock,
        )
        return response.to_dict()

    @staticmethod
    def serialize_list(instances: list[Infrastructure]) -> list[dict[str, Any]]:
        return [InfrastructureSerializer.serialize_instance(inst) for inst in instances]

File: launchpad-frontend/types/infrastructure.ts
export type InfrastructureStatus = 'PENDING' | 'PROVISIONING' | 'ACTIVE' | 'ERROR' | 'DESTROYING' | 'DESTROYED';

export type ComputeType = 'ecs_fargate' | 'eks';

export interface InvitedUserSummary {
  id: string;
  email: string;
  user_name: string;
  role: string;
}

export interface Infrastructure {
  id: string;
  name: string;
  cloud_provider: 'AWS';
  max_cpu: number;
  max_memory: number;
  code: string;
  compute_type: ComputeType;
  owner_id: string;
  user_id: string;
  status: InfrastructureStatus;
  is_cloud_authenticated: boolean;
  is_mock?: boolean;
  invited_users?: InvitedUserSummary[];
  metadata?: { aws_region?: string; [key: string]: string | undefined };
  created_at: string;
  updated_at: string;
  environment?: Environment;
}

export interface Environment {
  id: string;
  infrastructure_id: string;
  vpc_id?: string;
  ecs_cluster_arn?: string;
  alb_arn?: string;
  alb_dns?: string;
  ecr_repository_url?: string;
  task_execution_role_arn?: string;
  subnet_ids?: string[];
  security_group_ids?: string[];
  status: InfrastructureStatus;
}

export interface InfrastructureCreate {
  name: string;
  cloud_provider: 'aws';
  max_cpu: number;
  max_memory: number;
  code: string;
  compute_type: ComputeType;
  metadata?: { aws_region?: string;[key: string]: string | undefined };
}

// Returned only by POST /api/infrastructures/. The plaintext nonce is shown once and never re-served.
export interface InfrastructureCreateResponse extends Infrastructure {
  onboarding_token: string;
}
